src/collections/componentSetBuilder.ts

```diff
@@ -0,0 +1,171 @@
+        componentSet ??= new ComponentSet();
+
+        // Build a Set of metadata entries
+        metadata.metadataEntries.forEach((rawEntry) => {
+          const splitEntry = rawEntry.split(':').map((entry) => entry.trim());
+          // The registry will throw if it doesn't know what this type is.
+          registry.getTypeByName(splitEntry[0]);
+          const entry = {
+            type: splitEntry[0],
+            fullName: splitEntry.length === 1 ? '*' : splitEntry[1],
+          };
+          // Add to the filtered ComponentSet for resolved source paths,
+          // and the unfiltered ComponentSet to build the correct manifest.
+          compSetFilter.add(entry);
+          componentSet.add(entry);
+        });
+
+        const directoryPaths = options.metadata.directoryPaths;
+        logger.debug(`Searching for matching metadata in directories: ${directoryPaths.join(', ')}`);
+        const resolvedComponents = ComponentSet.fromSource({ fsPaths: directoryPaths, include: compSetFilter });
+        componentSet.forceIgnoredPaths = resolvedComponents.forceIgnoredPaths;
+        for (const comp of resolvedComponents) {
+          componentSet.add(comp);
+        }
+      }
+
+      // Resolve metadata entries with an org connection
+      if (org) {
+        logger.debug(`Building ComponentSet from targetUsername: ${org.username}`);
+        componentSet = await ComponentSet.fromConnection({
+          // eslint-disable-next-line @typescript-eslint/no-unsafe-call,@typescript-eslint/no-unsafe-member-access,@typescript-eslint/no-unsafe-assignment
+          usernameOrConnection: (await Aliases.fetch(org.username)) || org.username,
+          // exclude components based on the results of componentFilter function
+          // components with namespacePrefix where org.exclude includes manageableState (to exclude managed packages)
+          // components with namespacePrefix where manageableState equals undefined (to exclude components e.g. InstalledPackage)
+          // components where org.exclude includes manageableState (to exclude packages without namespacePrefix e.g. unlocked packages)
+          componentFilter: (component): boolean => !(org.exclude && org.exclude.includes(component?.manageableState)),
+        });
+      }
+    } catch (e) {
+      if ((e as Error).message.includes('Missing metadata type definition in registry for id')) {
+        // to remain generic to catch missing metadata types regardless of parameters, split on '
+        // example message : Missing metadata type definition in registry for id 'NonExistentType'
+        const issueType = (e as Error).message.split("'")[1];
+        throw new SfdxError(`The specified metadata type is unsupported: [${issueType}]`);
+      } else {
+        throw e;
+      }
+    }
+
+    // This is only for debug output of matched files based on the command flags.
+    // It will log up to 20 file matches.
+    if (logger.debugEnabled && componentSet.size) {
```

Comment:
I've noticed that `logger.debugEnabled` doesn't seem to be working properly.  Probably a bug in sfdx-core.